File: AyakaRail-master/pkg/http_server/handlers/dispatch.py
from fastapi import APIRouter
from fastapi.responses import PlainTextResponse
from pkg.config import (
    GAME_SERVER_HOST,
    GAME_SERVER_PORT,
    LUA_URL,
    LUA_VERSION,
    EX_RESOURCE_URL,
    ASSET_BUNDLE_URL,
    HTTP_SERVER_HOST,
    HTTP_SERVER_PORT,
)
from pkg.rail_proto.lib import Gateserver, ServerData, GlobalDispatchData
from base64 import b64encode

r = APIRouter()


@r.get("/query_dispatch", response_class=PlainTextResponse)
async def on_query_dispatch():
    rsp = bytes(
        GlobalDispatchData(
            msg="OK",
            top_sever_region_name="kamisato",
            server_list=[
                ServerData(
                    name="kamisato",
                    display_name="kamisato",
                    title="kamisato",
                    env_type="2",
                    msg="OK",
                    dispatch_url=f"http://{HTTP_SERVER_HOST}:{HTTP_SERVER_PORT}/query_gateway",
                )
            ],
        )
    )

    return b64encode(rsp)


@r.get("/query_gateway", response_class=PlainTextResponse)
async def on_query_gateway():
    rsp = bytes(
        Gateserver(
            use_tcp=False,
            ip=GAME_SERVER_HOST,
            port=GAME_SERVER_PORT,
            lua_url=LUA_URL,
            lua_version=LUA_VERSION,
            ex_resource_url=EX_RESOURCE_URL,
            asset_bundle_url=ASSET_BUNDLE_URL,
            unk1=True,
            unk2=True,
            unk3=True,
            unk4=True,
            unk5=True,
            unk6=True,
            unk7=True,
            unk9=True,
            unk10=True,
            unk11=True,
            unk12=True,
            unk13=True,
            unk14=True,
            unk15=True,
        )
    )

    return b64encode(rsp)

File: YoSR-roseate/source/sdk_server/handlers/dispatch.ml
module Tiny = Tiny_httpd
open Protocol


let query_dispatch _ =
  let server_data = 
    default_server_data
      ~name:"roseate"
      ~display_name:"roseate"
      ~dispatch_url:Config.dispatch_url
      ~env_type:"2"
      ~title:"roseate"
      ~msg:"OK"
    () in

  let global_dispatch_data = 
    default_global_dispatch_data
      ~msg:"OK"
      ~server_list:[server_data]
    () in

  let encoder = Pbrt.Encoder.create () in
  encode_pb_global_dispatch_data global_dispatch_data encoder;

  let serialized_data = Pbrt.Encoder.to_string encoder in
  let base64_encoded_data = Base64.encode_string serialized_data in
  Tiny.Response.make_string (Ok base64_encoded_data)


let query_gateway _ =
  let gateserver = 
    default_gateserver
      ~ip:Config.game_server_host
      ~port:Config.game_server_port
      ~asset_bundle_url:Config.asset_bundle_url
      ~ex_resource_url:Config.ex_resource_url
      ~lua_url:Config.lua_url
      ~lua_version:Config.lua_version
      ~use_tcp:true
      ~unk1:true
      ~unk2:true
      ~unk3:true
      ~unk4:true
      ~unk5:true
      ~unk6:true
      ~unk7:true
    () in

  let encoder = Pbrt.Encoder.create () in
  encode_pb_gateserver gateserver encoder;

  let serialized_data = Pbrt.Encoder.to_string encoder in
  let base64_encoded_data = Base64.encode_string serialized_data in
  Tiny.Response.make_string (Ok base64_encoded_data)
